streams/src/main/java/org/apache/kafka/streams/state/internals/buffer/InMemoryTimeOrderedKeyValueBufferStore.java

```diff
@@ -0,0 +1,219 @@
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals.buffer;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.header.Headers;
+import org.apache.kafka.common.header.internals.RecordHeaders;
+import org.apache.kafka.common.serialization.ByteArraySerializer;
+import org.apache.kafka.common.serialization.BytesSerializer;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.internals.ProcessorRecordContext;
+import org.apache.kafka.streams.processor.internals.ProcessorStateManager;
+import org.apache.kafka.streams.processor.internals.RecordBatchingStateRestoreCallback;
+import org.apache.kafka.streams.processor.internals.RecordCollector;
+import org.apache.kafka.streams.state.StoreBuilder;
+import org.apache.kafka.streams.state.internals.ContextualRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.ByteBuffer;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.Map;
+
+public class InMemoryTimeOrderedKeyValueBufferStore implements StateStore {
+    private static final Logger LOG = LoggerFactory.getLogger(InMemoryTimeOrderedKeyValueBufferStore.class);
+    private static final String BUFFER_TIME_HEADER = "_bufferTime";
+    private static final BytesSerializer KEY_SERIALIZER = new BytesSerializer();
+    private static final ByteArraySerializer VALUE_SERIALIZER = new ByteArraySerializer();
+
+    private final InMemoryTimeOrderedKeyValueBuffer buffer;
+    private final String storeName;
+
+    private RecordCollector collector;
+    private String changelogTopic;
+    private boolean open;
+
+    public static class Builder implements StoreBuilder<StateStore> {
+        private final String storeName;
+
+        public Builder(final String storeName) {
+            this.storeName = storeName;
+        }
+
+        @Override
+        public StoreBuilder<StateStore> withCachingEnabled() {
+            return null;
+        }
+
+        @Override
+        public StoreBuilder<StateStore> withCachingDisabled() {
+            return null;
+        }
+
+        @Override
+        public StoreBuilder<StateStore> withLoggingEnabled(final Map<String, String> config) {
+            return null;
```

Comment:
But I'm less certain about this... Should there be some way to disable logging? Right now, `init` just disables the changelog automatically if getting the record collector is an unsupported operation. This happens during unit tests.

I'm thinking more along the lines of people running code. Since the buffer is only in memory, it's guaranteed to lose records if you restart, unless there is a changelog there. As opposed to a Rocks K/V store, where you can make a reasonable gamble that your state will survive a restart (but maybe not a crash).

WDYT?

Reply:
Compare my comment above. As long as store is not plugable for suppress this question seems to be void.

Thus, if we do follow up KIP with public `BufferStoreInterface` we should allow to disable change logging IMHO. But this seems to be future work and I would exclude any discussion about it from this PR.